File: src/main/scala/com/johnsnowlabs/nlp/TableAssembler.scala
package com.johnsnowlabs.nlp

import com.johnsnowlabs.nlp.AnnotatorType.{DOCUMENT, TABLE}
import com.johnsnowlabs.nlp.annotators.common.TableData
import org.apache.spark.ml.param.{BooleanParam, Param}
import org.apache.spark.ml.util.{DefaultParamsReadable, Identifiable}

/** This transformer parses text into tabular representation. The input consists of DOCUMENT
  * annotations and the output are TABLE annotations. The source format can be either JSON or CSV.
  * The format of the JSON files should be:
  *
  * {{{
  * {
  *   "header": [col1, col2, ..., colN],
  *   "rows": [
  *     [val11, val12, ..., val1N],
  *     [val22, va22, ..., val2N],
  *     ...
  *    ]
  * }
  * }}}
  * The CSV format support alternative delimiters (e.g. tab), as well as escaping delimiters by
  * surrounding cell values with double quotes. For example:
  *
  * {{{
  * column1, column2, "column with, comma"
  * value1, value2, value3
  * "escaped value", "value with, comma", "value with double ("") quote"
  * }}}
  *
  * The transformer stores tabular data internally as JSON. The default input format is also JSON.
  *
  * ==Example==
  * {{{
  * import spark.implicits._
  * import com.johnsnowlabs.nlp.DocumentAssembler
  * import org.apache.spark.ml.Pipeline
  *
  * val csvData =
  *   """
  *     |"name", "money", "age"
  *     |"Donald Trump", "$100,000,000", "75"
  *     |"Elon Musk", "$20,000,000,000,000", "55"
  *     |""".stripMargin.trim
  *
  * val data =Seq(csvData).toDF("csv")
  *
  * val documentAssembler = new DocumentAssembler()
  *   .setInputCol("csv")
  *   .setOutputCol("document")
  *
  * val tableAssembler = new TableAssembler()
  *   .setInputCols(Array("document"))
  *   .setOutputCol("table")
  *   .setInputFormat("csv")
  *
  * val pipeline = new Pipeline()
  *   .setStages(
  *     Array(documentAssembler, tableAssembler)
  *     ).fit(data)
  *
  * val result = pipeline.transform(data)
  * result
  *   .selectExpr("explode(table) AS table")
  *   .select("table.result", "table.metadata.input_format")
  *   .show(false)
  *
  * +--------------------------------------------+-------------+
  * |result                                      |input_format |
  * +--------------------------------------------+-------------+
  * |{                                           |csv          |
  * | "header": ["name","money","age"],          |             |
  * |  "rows":[                                  |             |
  * |   ["Donald Trump","$100,000,000","75"],    |             |
  * |   ["Elon Musk","$20,000,000,000,000","55"] |             |
  * |  ]                                         |             |
  * |}                                           |             |
  * +--------------------------------------------+-------------+
  * }}}
  * @param uid
  *   required uid for storing annotator to disk
  * @groupname anno Annotator types
  * @groupdesc anno
  *   Required input and expected output annotator types
  * @groupname Ungrouped Members
  * @groupname param Parameters
  * @groupname setParam Parameter setters
  * @groupname getParam Parameter getters
  * @groupname Ungrouped Members
  * @groupprio param  1
  * @groupprio anno  2
  * @groupprio Ungrouped 3
  * @groupprio setParam  4
  * @groupprio getParam  5
  * @groupdesc param
  *   A list of (hyper-)parameter keys this annotator can take. Users can set and get the
  *   parameter values through setters and getters, respectively.
  */
class TableAssembler(override val uid: String)
    extends AnnotatorModel[TokenAssembler]
    with HasSimpleAnnotate[TokenAssembler] {

  def this() = this(Identifiable.randomUID("TableAssembler"))

  /** Output annotator types: DOCUMENT
    *
    * @group anno
    */
  override val outputAnnotatorType: AnnotatorType = TABLE

  /** Input annotator types: DOCUMENT, TOKEN
    *
    * @group anno
    */
  override val inputAnnotatorTypes: Array[String] = Array(DOCUMENT)

  protected val INPUT_FORMATS = Array("json", "csv")

  val inputFormat = new Param[String](
    this,
    "inputFormat",
    "Input table format. Supported formats: %s".format(INPUT_FORMATS.mkString(", ")))

  def setInputFormat(value: String): this.type = {
    require(
      INPUT_FORMATS.contains(value),
      "Invalid input format. Currently supported formats are: " + INPUT_FORMATS.mkString(", "))
    set(this.inputFormat, value.toLowerCase)
  }

  def getInputFormat: String = $(inputFormat)

  val csvDelimiter = new Param[String](this, "csvDelimiter", "CSV delimiter")

  def setCsvDelimiter(value: String): this.type = {
    set(this.csvDelimiter, value)
  }

  def getCsvDelimiter: String = $(csvDelimiter)

  val escapeCsvDelimiter =
    new BooleanParam(
      this,
      "escapeCsvDelimiter",
      "Escape Csv delimiter by surrounding values with double quotes")

  def setEscapeCsvDelimiter(value: Boolean): this.type = {
    set(this.escapeCsvDelimiter, value)
  }

  def getEscapeCsvDelimiter: Boolean = $(escapeCsvDelimiter)

  setDefault(inputFormat -> "json", csvDelimiter -> ",", escapeCsvDelimiter -> true)

  override def annotate(annotations: Seq[Annotation]): Seq[Annotation] = {

    annotations
      .map { annotation =>
        val jsonTable = $(inputFormat) match {
          case "json" => TableData.fromJson(annotation.result).toJson
          case "csv" =>
            TableData.fromCsv(annotation.result, $(csvDelimiter), $(escapeCsvDelimiter)).toJson
        }
        new Annotation(
          annotatorType = TABLE,
          begin = 0,
          end = jsonTable.length,
          result = jsonTable,
          metadata = annotation.metadata ++ Map("input_format" -> $(inputFormat)))
      }

  }
}

/** This is the companion object of [[TableAssembler]]. Please refer to that class for the
  * documentation.
  */
object TableAssembler extends DefaultParamsReadable[DocumentAssembler]

File: python/sparknlp/base/table_assembler.py
#  Copyright 2017-2022 John Snow Labs
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
"""Contains classes for the TableAssembler."""
from sparknlp.common import *


class TableAssembler(AnnotatorModel):
    """This transformer parses text into tabular representation. The input consists of DOCUMENT annotations and the
    output are TABLE annotations. The source format can be either JSON or CSV. The CSV format support alternative
    delimiters (e.g. tab), as well as escaping delimiters by surrounding cell values with double quotes.

    The transformer stores tabular data internally as JSON. The default input format is also JSON.

    ====================== ======================
    Input Annotation types Output Annotation type
    ====================== ======================
    ``DOCUMENT``           ``TABLE``
    ====================== ======================

    Parameters
    ----------
        inputFormat
            The format of the source representation of the table ('json' or 'csv')

        csvDelimiter
            The delimiter used for parsing CSV files (defailt is comma)

        escapeCsvDelimiter
            Whether to escape Csv delimiter by surrounding values with double quotes

    Examples
    --------
    >>> import sparknlp
    >>> from sparknlp.base import *
    >>> from sparknlp.annotator import *
    >>> from pyspark.ml import Pipeline
    >>>
    >>> document_assembler = DocumentAssembler() \\
    ...     .setInputCol("table_csv") \\
    ...     .setOutputCol("document_table")
    >>> table_assembler = TableAssembler()\\
    >>>     .setInputFormat("csv")\\
    >>>     .setInputCols(["document_table"])\\
    >>>     .setOutputCol("table")
    >>>
    >>> csv_data = "\\n".join([
    >>> "name, money, age",
    >>> "Donald Trump, \"100,000,000\", 75",
    >>> "Elon Musk, \"20,000,000,000,000\", 55"])
    >>> data = spark.createDataFrame([[csv_data]]) \\
    ...    .toDF("table_csv")
    >>> pipeline = Pipeline().setStages([
    ...     document_assembler,
     ...     table_assembler
    ... ]).fit(data)
    >>> result = pipeline.transform(data)
    >>> result.select("table").show(truncate=False)
    +-----------------------------------------------+
    |table                                          |
    +-----------------------------------------------+
    |[[table, 0, 118, {                             |
    |   "header":["name","money","age"],            |
    |   "rows":[                                    |
    |    ["Donald Trump","100,000,000","75"],       |
    |    ["Elon Musk","20,000,000,000,000","55"]]   |
    | },                                            |
    | [sentence -> 0, input_format -> csv], []]]    |
    +-----------------------------------------------+
"""
    name = "TableAssembler"

    inputAnnotatorTypes = [AnnotatorType.DOCUMENT]

    outputAnnotatorType = AnnotatorType.TABLE

    inputFormat = Param(
        Params._dummy(),
        "inputFormat",
        "Input format ('json' or 'csv')",
        typeConverter=TypeConverters.toString)

    csvDelimiter = Param(
        Params._dummy(),
        "csvDelimiter",
        "CSV delimiter",
        typeConverter=TypeConverters.toString)

    escapeCsvDelimiter = Param(
        Params._dummy(),
        "escapeCsvDelimiter",
        "Escape Csv delimiter by surrounding values with double quotes",
        typeConverter=TypeConverters.toBoolean)

    def setInputFormat(self, value):
        """Sets the table input format. The following formats are currently supported: json, csv.

        Parameters
        ----------
        value : str
            Table input format
        """
        return self._set(inputFormat=value)

    def setCsvDelimiter(self, value):
        """Sets the CSV delimiter.

        Parameters
        ----------
        value : str
            CSV delimiter
        """
        return self._set(csvDelimiter=value)

    def setEscapeCsvDelimiter(self, value):
        """Escape Csv delimiter by surrounding values with double quotes

        Parameters
        ----------
        value : bool
            True of Csv delimiter is escaped by surrounding values with double quotes
        """
        return self._set(escapeCsvDelimiter=value)

    @keyword_only
    def __init__(self, classname="com.johnsnowlabs.nlp.TableAssembler", java_model=None):
        super(TableAssembler, self).__init__(
            classname=classname,
            java_model=java_model
        )
        self._setDefault(
            inputFormat="json",
            csvDelimiter=",",
            escapeCsvDelimiter=True
        )
